doc/rst/technotes/optimization.rst

```diff
@@ -0,0 +1,662 @@
+generally lead to improvements in the scalability of your application on
+multiple nodes.
+
+To run locally oversubscribed, build Chapel with GASNet with the UDP
+conduit. Note that GASNet with the UDP conduit is not a high-performing
+configuration; in this setting that is OK because the communication
+overhead of UDP will approximate the communication overhead on a real
+multi-node system. You can use these commands to build with the UDP
+conduit:
+
+  .. code-block:: bash
+
+    export CHPL_COMM=gasnet
+    export CHPL_COMM_SUBSTRATE=udp
+    make
+
+Then, run with these settings:
+
+  .. code-block:: bash
+
+    export CHPL_RT_OVERSUBSCRIBED=yes
+    export GASNET_SPAWNFN=L
+    export CHPL_RT_MASTERIP=127.0.0.1
+
+You might also wish to run with a lower number of threads per simulated
+locale in order to reduce interference between the processes. For
+example, if you are working on a system with 32 cores, you might run
+with ``CHPL_RT_NUM_THREADS_PER_LOCALE=8 ./application -nl 4``. That
+uses 8 threads per locale, with 4 locales, so a total of 32 threads.
+
+See also :ref:`num-threads-per-locale`,
+:ref:`oversubscribed-execution`, and :ref:`using-udp`.
+
+
+Measuring Performance
+~~~~~~~~~~~~~~~~~~~~~
+
+Reproducible Performance Tests
+  It's important to have performance tests that are easy to run that you
+  can use to measure your program's performance. These performance test
+  cases need to be easy to repeat. Usually, that means setting aside some
+  problem configuration and input data to use as the main performance
+  test case. In some cases, you will need a battery of performance
+  tests instead of a single test case.
+
+Timing Regions
+  A common practice is to use :chpl:record:`~Time.stopwatch` to time
+  different regions of your program.  This coarse-grained timing
+  information indicates where your program is spending the most time. The
+  longest-running regions are the places where optimization will give the
+  most improvement to the program's performance. You can print out the
+  performance of each region after it is timed. It's a good idea to use a
+  ``config const`` to make the timing only print when you are studying
+  performance.
+
+Counting Communication Events
+  If you need to do most of your work by running locally oversubscribed
+  (see above) then it's particularly useful to measure performance by
+  counting the number of communication events. Counting communication
+  events can also be useful as a secondary performance metric for
+  multi-node runs. Reducing the number of communication events is a useful
+  strategy for optimizing communication.
+
+  You can use the :chpl:mod:`CommDiagnostics` module and
+  ``startCommDiagnostics()`` followed by ``getCommDiagnostics()`` to
+  count communication events. It's common to
+  ``writeln(getCommDiagnostics()``.  Note that the communication count
+  information will be easier to understand if you compile with
+  ``--no-cache-remote``.
```

Comment:
"... Because on-by-default remote caching can lead to surprising communication counts", or generally an explanation of why to use this flag, would be nice here.